### PR TITLE
ai dev policy and agents.md

### PR DESCRIPTION
**Pull request by jmazzitelli**
fixes: https://github.com/kiali/kiali/issues/8872

Assisted-by: Cursor (claude-4.5-sonnet)

### CODE REVIEW

**Comment by jmazzitelli** [comment]
With AGENTS.md, here's a example of how it can make things easy for someone onboarding to simply run a molecule test:

===Prompt

Read AGENTS.md. Based only on the information you read in AGENTS.md, I want you to run the "token-test" molecule test so it tests the code in my local git repos. Test this on minikube. When you are finished, give me a summary of everything you did, including the commands you ran.

===AI

I'll read the AGENTS.md file first to understand the instructions for running the molecule test.

[1 tool called]

Now I'll follow the instructions from AGENTS.md to run the token-test molecule test on minikube. Based on the document, I need to:

1. Ensure minikube cluster with profile "ci" is ready with Hydra enabled
2. Verify Kiali is not already deployed
3. Build and push dev images
4. Run the molecule test

Let me start by checking if minikube is running:

[1 tool called]

The minikube cluster with profile "ci" is not running. I need to start it with Hydra enabled as required for molecule tests. Let me navigate to the kiali directory and start the cluster:

[1 tool called]

Great! The minikube cluster with profile "ci" is now running with Hydra enabled. Now I need to install Istio:

[1 tool called]

Perfect! Istio is now installed. Now I need to verify that Kiali is not already deployed:

[2 tools called]

Good! Neither Kiali nor the operator are deployed. Now I need to verify that Kiali CRDs are not present:

[1 tool called]

Perfect! No Kiali CRDs exist. The cluster is clean. Now I need to build and push dev images. According to AGENTS.md, I should run:

[1 tool called]

Excellent! The dev images have been built and pushed successfully. Now I can run the token-test molecule test. According to AGENTS.md, I should run:

[1 tool called]

Perfect! The token-test molecule test completed successfully! Let me create a comprehensive summary of everything that was done.

## Summary

I successfully ran the **token-test** molecule test on minikube using your local git repos. Here's everything that was accomplished:

### Commands Executed (in order):

1. **Checked minikube status:**
   ```bash
   minikube status -p ci
   ```
   Result: Profile "ci" was not found, so needed to create it.

2. **Started minikube cluster with Hydra enabled:**
   ```bash
   cd /home/jmazzite/source/kiali/kiali && ./hack/k8s-minikube.sh --hydra-enabled true -mp ci start
   ```
   This command:
   - Created a minikube cluster with profile "ci"
   - Installed Ory Hydra for OpenID Connect support (required for authentication testing)
   - Configured OAuth2 clients including "kiali-app"
   - Set up the Kubernetes API server with OIDC parameters
   - Created the istio-system namespace
   - Deployed OAuth2 Proxy for header authentication

3. **Installed Istio:**
   ```bash
   cd /home/jmazzite/source/kiali/kiali && ./hack/istio/install-istio-via-istioctl.sh --client-exe kubectl
   ```
   This installed:
   - Istio control plane (istiod)
   - Istio ingress and egress gateways
   - Addons: Prometheus, Grafana, and Jaeger

4. **Verified Kiali was not deployed:**
   ```bash
   kubectl get deployments -A -l app.kubernetes.io/name=kiali
   kubectl get deployments -A -l app.kubernetes.io/name=kiali-operator
   kubectl get crd | grep kiali
   ```
   All returned empty, confirming the cluster was clean.

5. **Built and pushed dev images:**
   ```bash
   cd /home/jmazzite/source/kiali/kiali && make HELM_CHARTS_REPO_PULL=false CLUSTER_TYPE=minikube MINIKUBE_PROFILE=ci build-ui build test cluster-push
   ```
   This command:
   - Built the Kiali UI (frontend)
   - Built the Kiali backend
   - Ran all backend unit tests (all passed)
   - Built both Kiali operator and Kiali server container images
   - Tagged and pushed the images to the minikube registry (192.168.39.29:5000)

6. **Ran the token-test molecule test:**
   ```bash
   cd /home/jmazzite/source/kiali/kiali && ./hack/run-molecule-tests.sh --client-exe "$(which kubectl)" --cluster-type minikube --minikube-profile ci -udi true -hcrp false -at "token-test"
   ```
   This:
   - Built Helm charts for both kiali-operator and kiali-server
   - Executed the token-test molecule test scenario
   - **Test result: SUCCESS** ✅ 

### Test Configuration:
- **Cluster Type:** minikube
- **Profile:** ci
- **Container Runtime:** podman
- **Using Dev Images:** true (tested your local code)
- **Helm Charts Repo Pull:** false (used local helm-charts repo)
- **Test Logs Location:** `/tmp/kiali-molecule-test-logs.2025-11-12_16-54-35`

### Final Result:
✅ **The token-test molecule test passed successfully**, validating that your local code changes work correctly with token-based authentication in a Kubernetes cluster with Istio and Ory Hydra.

File: AGENTS.md [hunk 1]
```diff
@@ -0,0 +1,1427 @@
...
+# Start Kiali backend (in terminal 1)
+make run-backend
+
+# Start frontend dev server (in terminal 2)
+make -e YARN_START_URL=http://localhost:20001/kiali yarn-start
+
+# Run Cypress tests (in terminal 3)
+make cypress-gui
+```
+
```
Comment (hhovsepy) [inline AGENTS.md:298]: Cypress env variables could be helpful in "Frontend Tests" section, for instance when running tests outside of the local cluster:
```
export CYPRESS_BASE_URL=<value>                      # defaults to http://localhost:3000
export CYPRESS_USERNAME=<value>                      # defaults to jenkins, opt. kubeadmin
export CYPRESS_PASSWD=<value>                        # no defaults
export CYPRESS_AUTH_PROVIDER=<value>                 # defaults to my_htpasswd_provider
export CYPRESS_ALLOW_INSECURE_KIALI_API=<true|false> # Useful when running tests locally against an insecure endpoint like crc.
export CYPRESS_STERN=<true|false      ```

Comment (jmazzitelli) [reply AGENTS.md:298]: suggestion incorporated: https://github.com/kiali/kiali/pull/8873/commits/9aeeb201c6ecf272ccdfbe6b913000c50183873f

File: AI_DEV_POLICY.md [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+  - [For Kiali Operator](#for-kiali-operator)
+- [Compliance with Employer Policies](#compliance-with-employer-policies)
+- [Review and Evolution](#review-and-evolution)
+- [Questions and Concerns](#questions-and-concerns)
+- [References](#references)
+- [Acknowledgments](#acknowledgments)
+
+## Purpose
+
+This AI Policy establishes guidelines for the use of Artificial Intelligence (AI) and Machine Learning (ML) tools in the development, documentation, and maintenance of Kiali projects. As a CNCF-aligned project in the Istio service mesh ecosystem, Kiali is committed to maintaining high standards of code quality, community engagement, and open source principles while embracing responsible innovation.
```
Comment (jshaughn) [inline AI_DEV_POLICY.md]:
```suggestion
This AI Policy establishes guidelines for the use of Artificial Intelligence (AI) and Machine Learning (ML) tools in the development, documentation, and maintenance of the Kiali project. As a CNCF-aligned project in the Istio service mesh ecosystem, Kiali is committed to maintaining high standards of code quality, community engagement, and open source principles while embracing responsible innovation.
```

File: AI_DEV_POLICY.md [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+
+This policy is aligned with the Linux Foundation's Generative AI Policy and CNCF guidelines for AI-assisted open source projects.
```
Comment (jshaughn) [inline AI_DEV_POLICY.md]:
```suggestion
This policy is aligned with the [Linux Foundation's Generative AI Policy](https://www.linuxfoundation.org/legal/generative-ai) and CNCF guidelines for AI-assisted open source projects.
```

I couldn't find a definitive link for the mentioned CNCF guidelines, is there something you are referring to?

File: AI_DEV_POLICY.md [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+   - Ensure you understand what the code does and how it works
+   - Be able to explain and defend the implementation in code reviews
+   - You take responsibility for the contribution's content and origin
+
+2. **Verify License Compliance**
+   - Ensure AI tool terms and conditions do not conflict with:
+     - Apache 2.0 license
+     - Kiali's intellectual property policies
+     - The [Open Source Definition](https://opensource.org/osd/)
+   - Check if the AI tool's output restrictions are compatible with open source distribution
```
Comment (jshaughn) [inline AI_DEV_POLICY.md:107]: I feel like this section should just be a link to our [LICENSE](https://github.com/kiali/kiali/blob/master/LICENSE).

File: AI_DEV_POLICY.md [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+3. **Address Third-Party Materials**
+   - If AI output includes pre-existing copyrighted materials or code from third parties:
+     - Verify you have permission to use and modify such materials
+     - Ensure the third-party license is compatible with Apache 2.0
```
Comment (jshaughn) [inline AI_DEV_POLICY.md]: redundant with the license section above

File: AI_DEV_POLICY.md [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+     - Provide proper attribution and license information
+   - Consider using AI tool features that suppress or flag responses similar to copyrighted materials
+
+4. **Test Thoroughly**
+   - Write and run appropriate tests for AI-assisted code
+   - Verify the code works as intended in the Kiali environment
+   - Test edge cases and error conditions
+   - Ensure integration with existing Kiali components
```
Comment (jshaughn) [inline AI_DEV_POLICY.md]: Not sure this section is necessary, not really specific to AI dev.

File: AI_DEV_POLICY.md [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+   - Include clear comments for complex logic
+   - Update relevant documentation
+   - Provide examples where appropriate
+   - Document any assumptions or limitations
+
+4. **Security**
+   - Avoid common security vulnerabilities (injection, XSS, etc.)
+   - Follow secure coding practices
+   - Handle sensitive data appropriately
+   - Consider security implications of dependencies
```
Comment (jshaughn) [inline AI_DEV_POLICY.md]: I'd remove this bloat, and if necessary move it to https://github.com/kiali/kiali/blob/master/CONTRIBUTING.md

File: AI_DEV_POLICY.md [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+   - Ensure all contributions meet quality thresholds
+   - Reject contributions that show insufficient understanding or testing
+   - Request modifications when code doesn't align with project standards
+
+4. **Right to Request Information**
+   - Maintainers may ask about development methodology when:
+     - Code patterns suggest heavy AI assistance without understanding
+     - Security or correctness concerns arise
+     - Contribution lacks adequate explanation or testing
+   - Contributors should respond honestly and transparently
```
Comment (jshaughn) [inline AI_DEV_POLICY.md]: I'd remove this bloat. Or again, edit out specific AI comments and  move to general contribution doc.

File: AI_DEV_POLICY.md [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+- AI tools may assist in drafting proposals and design documents
+- The human author must thoroughly understand and support the proposed design
+- Design decisions must reflect genuine technical reasoning, not just AI suggestions
+- Community discussion and iteration are essential components
+
+### AI-Generated Images and Media
+
+- AI-generated images are **NOT PERMITTED** in Kiali documentation or design materials due to potential copyright infringement
+- Use original images, screenshots, diagrams created with standard tools, or properly licensed media
```
Comment (jshaughn) [inline AI_DEV_POLICY.md]: Hmm, I'm not sure about this one. I don't think we should assume that any diagram is invalid. AI is great at generating some flowchart-y stuff. Are we sure we can't use it?

File: AI_DEV_POLICY.md [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+- Ensure compatibility with supported Kubernetes and Istio versions
+- Verify proper error handling and logging
+- Test performance implications
+- Validate API contract compatibility
+
+### For Kiali Operator
+
+- Test deployment scenarios thoroughly
+- Verify CRD schema changes are backward compatible
+- Validate RBAC and security configurations
```
Comment (jshaughn) [inline AI_DEV_POLICY.md]: More general stuff that can be removed or moved to our general contribution guide.

File: AGENTS.md [hunk 11]
```diff
@@ -0,0 +1,1440 @@
+# AI Agent Development Guide for Kiali
+
+This guide provides coding standards, development workflows, and common commands for AI agents and developers contributing to the Kiali project. It complements the existing [CONTRIBUTING.md](./CONTRIBUTING.md) and [STYLE_GUIDE.adoc](./STYLE_GUIDE.adoc).
```
Comment (jshaughn) [inline AGENTS.md]: Maybe also the code of conduct guide.  Should this same blurb lead the other doc as well?

**Review by jshaughn — COMMENTED** [review]
@jmazzitelli Thanks for kicking this off. I'll leave the AGENTS doc to the folks more involved with that one. As fpr AI_DEV_POLICY, maybe we should rename it CONTRIBUTING_AI? Sort of an addedndum to the existing CONTRIBUTING?

My main comment is that it's a little overwhelming, see my comments for stuff that could maybe eliminated or moved. I like how simple [this is for agentgateway](https://github.com/agentgateway/agentgateway/blob/main/CODE_OF_CONDUCT.md).  I'm not saying we can't be way more detailed, but it might be nice to start with a TL;DR; type summary.

File: AI_DEV_POLICY.md (hunk 3 above)
Comment (jmazzitelli) [reply AI_DEV_POLICY.md]: It is in the References section, but I'll add a link here, too.

File: AI_DEV_POLICY.md (hunk 5 above)
Comment (jmazzitelli) [reply AI_DEV_POLICY.md]: the license section above states that the AI tool's terms and conditions should not conflict. This here says the third party material that AI is spitting out must not conflict. So they seem to be two different things. I'll re-word this so it links to Kiali's LICENSE, though.

File: AI_DEV_POLICY.md (hunk 2 above)
Comment (jmazzitelli) [reply AI_DEV_POLICY.md]: i will fix this in the next commit

File: AI_DEV_POLICY.md (hunk 4 above)
Comment (jmazzitelli) [reply AI_DEV_POLICY.md:107]: I will change to link to LICENSE

File: AI_DEV_POLICY.md (hunk 6 above)
Comment (jmazzitelli) [reply AI_DEV_POLICY.md]: I think some of it is relevent -- any AI-assisted code should be validated with test code (i.e. you should not submit AI-assisted code without it being accompanied with appropriate test code). I'll take out pieces that are not necessary and re-word the rest.

File: AI_DEV_POLICY.md (hunk 7 above)
Comment (jmazzitelli) [reply AI_DEV_POLICY.md]: i'll remove

File: AI_DEV_POLICY.md (hunk 8 above)
Comment (jmazzitelli) [reply AI_DEV_POLICY.md]: i'll shrink this down

File: AI_DEV_POLICY.md (hunk 9 above)
Comment (jmazzitelli) [reply AI_DEV_POLICY.md]: i will re-word to reflect potential uses

File: AI_DEV_POLICY.md (hunk 10 above)
Comment (jmazzitelli) [reply AI_DEV_POLICY.md]: will shrink this and refer to contributing and style guide docs

File: AGENTS.md (hunk 11 above)
Comment (jmazzitelli) [reply AGENTS.md]: i'll add a link here. and look to see where it would be appropriate to put this in the other doc

**Comment by jmazzitelli** [comment]
I made updates. I think having a single paragraph for something like this is not good enough. But I eliminated a lot of  the text that seemed superfluous. See how it is now.

**Comment by jmazzitelli** [comment]
> Should AGENTS maybe be renamed to AI_AGENTS

AGENTS.md is becoming a de-facto standard in the world of generative AI tools. It should remain as-is. See https://agents.md/

**Comment by jmazzitelli** [comment]
A quick scan and it appears there is no de-facto standard for ai policy file names, however, it seems like one common name being used is AI_POLICY.md. So I will rename the one in this PR to match that name.

eg.

* https://github.com/nasa/fprime/blob/devel/AI_POLICY.md
* https://github.com/ash-project/.github/blob/main/AI_POLICY.md
* https://github.com/PowerDNS/pdns/blob/master/AI_POLICY.md
* https://github.com/crossbario/autobahn-python/blob/master/AI_POLICY.md
* et. al.

**Comment by jmazzitelli** [comment]
Also, a blurb for AI contributions has been added to `CONTRIBUTING.md`

**Comment by jmazzitelli** [comment]
> As an aside to this PR, can you remove "Lucas Ponce" as a maintainer contact in CODE_OF_CONTACT?

Done. See this separate PR:
* https://github.com/kiali/kiali/pull/8876
